# Tutorial for placement of new sequences into our *Nostoc* phylogenetic framework

The phylogenomic species tree (wASTRAL) and the *rbcLX* section trees that we inferred as part of the paper are available on the [T-BAS](https://guide-tbas.cifr.ncsu.edu/tbas) platform. T-BAS has a user-friendly interface to place unknown taxa into existing phylogenetic trees using phylogenetic placement algorithms or de-novo phylogenetic inference with or without backbone constraints. All you need is a file(s) with sequences from the markers included in T-BAS and it will add the unknown taxa to the alignments, perform the phylogenetic analyses, and display and interactive tree plot with the reference taxa and the queries.

T-BAS has its own tutorials for [phylogenetic placement and de-novo inference](https://tbas.cifr.ncsu.edu/tbas2_3/pages/tbas-tutorial_3.php) as well as for [viewing trees, alignments, and specimen metadata](https://tbas.cifr.ncsu.edu/tbas2_3/pages/tbas-tutorial_2.php). You can refer to those for general instructions on how to use the platform.

The purpose of this tutorial is to give guidelines for some specific use cases with unknown *Nostoc* sequences.

## Overview of the available datasets

When you go the the T-BAS trees page, you will see a guide tree with a summary of all the datasets available in the dataset organized by taxonomic ranks:

![guide_tree](https://github.com/cjpardodelahoz/nostoc/tree/main/tbas_tutorial/screen/guide_tree.png)



## Unknown queries with single locus data

